Content with -webkit-user-select: none should be skipped during copy

rniwa opened the pull request:
#### 166aef8967703338451b376ce492edc0fa79d86a
<pre>
Content with -webkit-user-select: none should be skipped during copy
<a href="https://bugs.webkit.org/show_bug.cgi?id=80159">https://bugs.webkit.org/show_bug.cgi?id=80159</a>

Reviewed by Darin Adler.

Exclude content with -webkit-user-select: none when copying. This patch introduces
TextIteratorBehavior::IgnoresUserSelectNone to exclude such content from plain text,
and adds IgnoreUserSelectNone as an argument to serializePreservingVisualAppearance
to exclude such content from HTML. It also updates HTMLConverter::_enterElement to
exclude such content from RTFD.

Finally, this patch also updates DOMSelection::toString to have the same treatment.

* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-content-with-user-select-none.html: Added.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectedText const):
(WebCore::Editor::selectedTextForDataTransfer const):

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::isRendererVisible): Added.
(WebCore::TextIterator::advance):
(WebCore::hasVisibleTextNode):
(WebCore::TextIterator::handleTextNodeFirstLetter):

* Source/WebCore/editing/TextIteratorBehavior.h:

* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::selectionInHTMLFormat):

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_enterElement):

* Source/WebCore/editing/gtk/EditorGtk.cpp:
(WebCore::Editor::writeSelectionToPasteboard):

* Source/WebCore/editing/libwpe/EditorLibWPE.cpp:
(WebCore::Editor::writeSelectionToPasteboard):

* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::writeImageToPasteboard):

* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::StyledMarkupAccumulator):
(WebCore::StyledMarkupAccumulator::renderedTextRespectingRange):
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):
(WebCore::serializePreservingVisualAppearanceInternal):
(WebCore::serializePreservingVisualAppearance):
(WebCore::sanitizedMarkupForFragmentInDocument):

* Source/WebCore/editing/markup.h:

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::createFromSelection):

* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::toString const):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(CopyRTF.StripsUserSelectNone):

Canonical link: <a href="https://commits.webkit.org/257749@main">https://commits.webkit.org/257749@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0694fc8efabaf436f47ce043db7e49bd2ad2a5b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99817 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8997 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32905 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109183 "Updated wpe dependencies (failure)") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169420 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103820 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9911 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86287 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92296 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107092 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105584 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90732 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34198 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22120 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2812 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23630 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46009 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8893 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43103 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2729 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4620 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
Seems like a good implementation. A little frustrating when there is text you can’t copy but I’ll assume this is good for users, not just interoperability and the concerns of website developers.

rniwa committed:
Content with -webkit-user-select: none should be skipped during copy
https://bugs.webkit.org/show_bug.cgi?id=80159

Reviewed by Darin Adler.

Exclude content with -webkit-user-select: none when copying. This patch introduces
TextIteratorBehavior::IgnoresUserSelectNone to exclude such content from plain text,
and adds IgnoreUserSelectNone as an argument to serializePreservingVisualAppearance
to exclude such content from HTML. It also updates HTMLConverter::_enterElement to
exclude such content from RTFD.

Finally, this patch also updates DOMSelection::toString to have the same treatment.

* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-content-with-user-select-none.html: Added.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectedText const):
(WebCore::Editor::selectedTextForDataTransfer const):

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::isRendererVisible): Added.
(WebCore::TextIterator::advance):
(WebCore::hasVisibleTextNode):
(WebCore::TextIterator::handleTextNodeFirstLetter):

* Source/WebCore/editing/TextIteratorBehavior.h:

* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::selectionInHTMLFormat):

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_enterElement):

* Source/WebCore/editing/gtk/EditorGtk.cpp:
(WebCore::Editor::writeSelectionToPasteboard):

* Source/WebCore/editing/libwpe/EditorLibWPE.cpp:
(WebCore::Editor::writeSelectionToPasteboard):

* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::writeImageToPasteboard):

* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::StyledMarkupAccumulator):
(WebCore::StyledMarkupAccumulator::renderedTextRespectingRange):
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):
(WebCore::serializePreservingVisualAppearanceInternal):
(WebCore::serializePreservingVisualAppearance):
(WebCore::sanitizedMarkupForFragmentInDocument):

* Source/WebCore/editing/markup.h:

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::createFromSelection):

* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::toString const):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(CopyRTF.StripsUserSelectNone):

Canonical link: https://commits.webkit.org/257749@main
